libbeat/publisher/queue/spool/inbroker.go

```diff
@@ -0,0 +1,482 @@
+	}
+
+	return true
+}
+
+// stateWithTimer is the brokers active state, if the write buffer is not empty.
+// The flush timer is enabled as long as the broker is in this state.
+// ACKs from the output are ignored, as events can still be added to the write
+// buffer.
+//
+// stateWithTimer transitions:
+//   -> stateWithTimer
+//        - if serializing failed
+//        - if event is added to buffer, without flush
+//        - flush, but more events are available in the buffer (might reset timer)
+//   -> stateEmpty if all events have been flushed
+//   -> stateBlocked if queue did return an error on write/flush (Flush failed)
+func (b *inBroker) stateWithTimer() bool {
+	log := b.ctx.logger
+
+	select {
+	case <-b.ctx.Done():
+		return false
+
+	case req := <-b.events:
+		log.Debug("inbroker (stateWithTimer): new event")
+
+		buf, st, err := b.encodeEvent(&req)
+		if err != nil {
+			log.Debug("  inbroker (stateWithTimer): encode failed")
+			b.respondDrop(&req)
+			break
+		}
+
+		count := b.bufferedEvents
+		err = b.addEvent(buf, st)
+		if err != nil {
+			log.Debug("  inbroker (stateWithTimer): append failed, blocking")
+			b.state = (*inBroker).stateBlocked
+			break
+		}
+
+		if b.bufferedEvents < count {
+			log.Debug("  inbroker (stateWithTimer): buffer flushed")
+
+			// write buffer has been flushed, reset timer and broker state
+			if b.bufferedEvents == 0 {
+				b.timer.Stop(false)
+				b.state = (*inBroker).stateEmpty
+			} else {
+				// restart timer, as new event is most likely the only event buffered
+				// -> reduce IO
+				log.Debug("  inbroker (stateWithTimer): start flush timer")
+				b.timer.Restart()
+			}
+		}
+
+	case req := <-b.pubCancel:
+		b.handleCancel(&req)
+
+	case <-b.timer.C:
+		log.Debug("inbroker (stateWithTimer): flush timeout")
+
+		b.timer.Stop(true)
+
+		err := b.flushBuffer()
+		if err != nil {
+			log.Debug("  inbroker (stateWithTimer): flush failed, blocking")
+			b.state = (*inBroker).stateBlocked
+			break
+		}
+
+		log.Debug("  inbroker (stateWithTimer): flush succeeded")
+
+		if b.bufferedEvents > 0 {
+			// flush did not push all events? Restart timer.
+			log.Debug("  inbroker (stateWithTimer): start flush timer")
+			b.timer.Start()
+			break
+		}
+
+		b.state = (*inBroker).stateEmpty
+
+	case <-b.sigACK:
+		// ignore ACKs as long as we can write without blocking
+	}
+
+	return true
+}
+
+// stateBlocked is the brokers active state if the write buffer can not accept
+// any new events.
+// The broker will wait for an ACK signal from the outputs and retry flushing,
+// in the hope of enough memory being available to flush the buffers.
+// If flush did succeed, we try to add the pending event.
+// For the time the broker is in this state, no events from any producers will
+// be accepted. Thusly all producers will block. Closing a producer, unblocks
+// the producer. The producers event (after close) might be processed or
+// ignored in the future.
+//
+// stateBlocked transitions:
```

Comment:
In the case of Filebeat this means the harvesters are blocked. In case of metricbeat does means the last events coming in will be dropped? So if the output becomes available again, all events in the queue will be sent and the events which came in between "queue full" and "output available again" will be lost.

Reply:
It's the same behavior as for the in memory queue. If the queue is full, it is full! If queue is full, it blocks (no event is dropped).

Block vs. drop (beside having similar effects on metricbeat) do have subtle differences. Metricbeat could report/monitor long block periods by having a watchdog checking metricsets actually report events with the configured period (or multiple of the configured periods).

It's by design. Administrator pre-allocate/assign disk space for use by beats and beats guarantees (best effort) it does not use any more disk space once the queue is full. Better have beats block if queue is full then making a system go out of space.